Fix major problems in the AsyncParser design.

non opened the pull request:
The async parser was introduced to handle a very specific use case:
parsing a (whitespace delimited) stream of relatively small (<1M) JSON
objects. Its design, buffering strategy, and strategy for resuming
parsing were all based on this scenario. In scenarios where the user
wanted to use many small chunks to parse a much larger value (i.e. 60M)
parser performance would degrate horribly compared to the synchronous
(all-at-once) parser.

Furthermore, the error-recovery strategy was based around similar
assumptions, and was not reliable in general (without frim commitments
from the user about delimiters and escaping it is hard to see how it
could be).

This commit does several things:
1. Fix the major performance problems seen when async parsing one
   large value. While this is still not an ideal situation,
   performance is now roughly on-par with the equivalent sync parser.
2. Remove the error-recovery option for the AsyncParser. It didn't
   work reliably and was not being relied on by anyone.
3. Create 3 different async parsing modes:
   
   a. Streaming: This is the existing behavior (returning a collection
      of whitespace-delimited JSON values).
   
   b. Json: This mode expects the input to be a single well-formed
      JSON value. In the future this should become the default
      behavior.
   
   c. Unwrapping: This mode is like (b) but if the JSON value is an
      array, the parser will instead parse the contents of the array
      as a stream and return those values, rather than the outer
      array.
   
   Mode (c) is especially exciting, since for situations where users
   need to parse very large arrays of values. Unlike the sync parser
   (or the async parser in modes (a) and (b)) unwrapping means that
   the parser itself does not need to accumulate and store all the
   parsed objects itself.
4. The commit adds some additional tests and benchmarks to measure the
   performance of these parsers.

None of the synchronous parsers should be affected by these changes.

non committed:
Fix major problems in the AsyncParser design.

The async parser was introduced to handle a very specific use case:
parsing a (whitespace delimited) stream of relatively small (<1M) JSON
objects. Its design, buffering strategy, and strategy for resuming
parsing were all based on this scenario. In scenarios where the user
wanted to use many small chunks to parse a much larger value (i.e. 60M)
parser performance would degrate horribly compared to the synchronous
(all-at-once) parser.

Furthermore, the error-recovery strategy was based around similar
assumptions, and was not reliable in general (without frim commitments
from the user about delimiters and escaping it is hard to see how it
could be).

This commit does several things:

 1. Fix the major performance problems seen when async parsing one
    large value. While this is still not an ideal situation,
    performance is now roughly on-par with the equivalent sync parser.

 2. Remove the error-recovery option for the AsyncParser. It didn't
    work reliably and was not being relied on by anyone.

 3. Create 3 different async parsing modes:

    a. Streaming: This is the existing behavior (returning a collection
       of whitespace-delimited JSON values).

    b. Json: This mode expects the input to be a single well-formed
       JSON value. In the future this should become the default
       behavior.

    c. Unwrapping: This mode is like (b) but if the JSON value is an
       array, the parser will instead parse the contents of the array
       as a stream and return those values, rather than the outer
       array.

    Mode (c) is especially exciting, since for situations where users
    need to parse very large arrays of values. Unlike the sync parser
    (or the async parser in modes (a) and (b)) unwrapping means that
    the parser itself does not need to accumulate and store all the
    parsed objects itself.

 4. The commit adds some additional tests and benchmarks to measure the
    performance of these parsers.

None of the synchronous parsers should be affected by these changes.

dcsobral commented on json/src/main/scala/blueeyes/json/AsyncParser.scala:
```diff
@@ -15,17 +16,59 @@ private[json] class AsyncException extends Exception
+  def apply(): AsyncParser = stream()
+
+  def stream(): AsyncParser = 
+    new AsyncParser(-1, 0, Nil, new Array[Byte](131072), 0, 131072, 0, false, 0)
```
When something has too many primitive types whose purpose isn't clear, I like to use named parameters to help whoever gets to read it later. It also helps avoiding problems due to parameter position changes.

dcsobral commented on json/src/main/scala/blueeyes/json/AsyncParser.scala:
```diff
@@ -15,17 +16,59 @@ private[json] class AsyncException extends Exception
+    new AsyncParser(-5, 0, Nil, new Array[Byte](131072), 0, 131072, 0, false, 1)
 }
 
+// We have a lot of fields here.
```
I think this would be better as a scaladoc, even if the constructor is protected. Protected members do show up on scaladoc, if you change visibility.

dcsobral commented on json/src/main/scala/blueeyes/json/ByteBasedParser.scala:
```diff
@@ -76,7 +76,7 @@ private[json] trait ByteBasedParser extends Parser {
         die(j, "invalid UTF-8 encoding")
       }
-      j = reset(j)
+      //j = reset(j)
```
At least explain why this is commented out instead of simply removed.

non replied:
I'll just remove it. Unfortunately I already pushed one clean up commit. I'll wait and push another when you've reviewed the whole thing.

non committed:
Some minor suggested clean up.

This commit takes dcsobral's suggestion of using named parameters for
AsyncParser's constructor. It also cleans up some other comments.

non committed:
more clean up based on dcsobral's suggestions
